data-prepper-plugins/flatten-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/flatten/FlattenProcessor.java

```diff
@@ -20,34 +20,47 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 @DataPrepperPlugin(name = "flatten", pluginType = Processor.class, pluginConfigurationType = FlattenProcessorConfig.class)
 public class FlattenProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
     private static final Logger LOG = LoggerFactory.getLogger(FlattenProcessor.class);
-
     private static final String SEPARATOR = "/";
     private final FlattenProcessorConfig config;
     private final ExpressionEvaluator expressionEvaluator;
+    private final Set<String> excludeKeySet = new HashSet<>();
 
     @DataPrepperPluginConstructor
     public FlattenProcessor(final PluginMetrics pluginMetrics, final FlattenProcessorConfig config, final ExpressionEvaluator expressionEvaluator) {
         super(pluginMetrics);
         this.config = config;
         this.expressionEvaluator = expressionEvaluator;
+        excludeKeySet.addAll(config.getExcludeKeys());
     }
 
     @Override
     public Collection<Record<Event>> doExecute(final Collection<Record<Event>> records) {
         for (final Record<Event> record : records) {
             final Event recordEvent = record.getData();
+            final Map<String, Object> excludeMap = new HashMap<>();
 
             try {
                 if (config.getFlattenWhen() != null && !expressionEvaluator.evaluateConditional(config.getFlattenWhen(), recordEvent)) {
                     continue;
                 }
 
+                // remove fields specified in "exclude_keys" from the event temporarily
+                for (final String key : excludeKeySet) {
+                    final String keyInEvent = getJsonPointer(config.getSource(), key);
```

Comment:
I think you should be able to refactor the collection of the excludeKeyJsonPointers outside the record iterator since it is independent of each record event.

Reply:
That's true. The json pointers don't have to be constructed for every events. Updated PR.